File: src/test/java/org/frawa/elmtest/core/ElmPluginHelperTest.java
package org.frawa.elmtest.core;

import com.intellij.psi.PsiElement;
import com.intellij.testFramework.ParsingTestCase;
import org.elm.ide.test.core.LabelUtils;
import org.elm.lang.core.parser.ElmParserDefinition;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;
import java.util.Arrays;

import static org.frawa.elmtest.core.ElmPluginHelper.getPsiElement;

public class ElmPluginHelperTest extends ParsingTestCase {

    public ElmPluginHelperTest() {
        super("elmPluginHelper", "elm", new ElmParserDefinition());
    }

    @Override
    protected String getTestDataPath() {
        return "src/test/resources";
    }

    @NotNull
    @Override
    // see file resources/elmPluginHelper/Navigation
    protected String getTestName(boolean lowercaseFirstLetter) {
        return "Navigation";
    }

    public void testTopLevelSuite() {
        doTest(false);
        assertSuite(27, "suite1");
    }

    public void testTestInSuite() {
        doTest(false);
        assertTest(55, "suite1", "test1");
    }

    public void testTopLevelTest() {
        doTest(false);
        assertTest(137, "test1");
    }

    public void testNestedSuitesAndTests() {
        doTest(false);
        assertSuite(207, "suite2");
        assertTest(235, "suite2", "test1");
        assertSuite(291, "suite2", "nested1");
        assertTest(324, "suite2", "nested1", "test1");
    }

    public void testMissingTopLevelSuite() {
        doTest(false);
        assertMissing("suiteMissing");
    }

    public void testMissingTopLevelTest() {
        doTest(false);
        assertMissing("testMissing");
    }

    public void testMissingNestedSuitesAndTests() {
        doTest(false);
        assertSuite(207, "suite2");
        assertMissing("suite2", "testMissing");
        assertFallback("suite2", "suite2", "nestedMissing");
        assertFallback("nested1", "suite2", "nested1", "testMissing");
    }

    private void assertSuite(int offset, String... labels) {
        Path path = LabelUtils.INSTANCE.toPath(Arrays.asList(labels));
        PsiElement element = getPsiElement(true, path.toString(), myFile);

        String expected = String.format("describe \"%s\"", labels[labels.length - 1]);
        assertEquals(expected, firstLine(text(element)));
        assertEquals(offset, element.getNode().getStartOffset());
    }

    private void assertTest(int offset, String... labels) {
        Path path = LabelUtils.INSTANCE.toPath(Arrays.asList(labels));
        PsiElement element = getPsiElement(false, path.toString(), myFile);

        String expected = String.format("test \"%s\"", labels[labels.length - 1]);
        assertEquals(expected, text(element));
        assertEquals(offset, element.getNode().getStartOffset());
    }

    private void assertMissing(String... labels) {
        Path path = LabelUtils.INSTANCE.toPath(Arrays.asList(labels));
        PsiElement element = getPsiElement(false, path.toString(), myFile);
        assertSame(myFile, element);
    }

    private void assertFallback(String fallback, String... labels) {
        Path path = LabelUtils.INSTANCE.toPath(Arrays.asList(labels));
        PsiElement element = getPsiElement(true, path.toString(), myFile);

        String expected = String.format("describe \"%s\"", fallback);
        assertEquals(expected, firstLine(text(element)));
    }

    @NotNull
    private String text(PsiElement element) {
        return element.getText().trim();
    }

    private String firstLine(String text) {
        return text.split("\n")[0];
    }

}

File: src/test/kotlin/org/elm/ide/test/core/ElmPluginHelperTest.kt
package org.elm.ide.test.core

import com.intellij.psi.PsiElement
import com.intellij.testFramework.ParsingTestCase
import org.elm.ide.test.core.ElmPluginHelper.getPsiElement
import org.elm.ide.test.core.LabelUtils.toPath
import org.elm.lang.core.parser.ElmParserDefinition

class ElmPluginHelperTest : ParsingTestCase("elmPluginHelper", "elm", ElmParserDefinition()) {

    override fun getTestDataPath(): String {
        return "src/test/resources"
    }

    override// see file resources/elmPluginHelper/Navigation
    fun getTestName(lowercaseFirstLetter: Boolean): String {
        return "Navigation"
    }

    fun testTopLevelSuite() {
        doTest(false)
        assertSuite(27, "suite1")
    }

    fun testTestInSuite() {
        doTest(false)
        assertTest(55, "suite1", "test1")
    }

    fun testTopLevelTest() {
        doTest(false)
        assertTest(137, "test1")
    }

    fun testNestedSuitesAndTests() {
        doTest(false)
        assertSuite(207, "suite2")
        assertTest(235, "suite2", "test1")
        assertSuite(291, "suite2", "nested1")
        assertTest(324, "suite2", "nested1", "test1")
    }

    fun testMissingTopLevelSuite() {
        doTest(false)
        assertMissing("suiteMissing")
    }

    fun testMissingTopLevelTest() {
        doTest(false)
        assertMissing("testMissing")
    }

    fun testMissingNestedSuitesAndTests() {
        doTest(false)
        assertSuite(207, "suite2")
        assertMissing("suite2", "testMissing")
        assertFallback("suite2", "suite2", "nestedMissing")
        assertFallback("nested1", "suite2", "nested1", "testMissing")
    }

    fun testFuzzTest() {
        doTest(false)
        assertFuzz(495, "fuzz1")
        assertFuzz(388, "suite2", "fuzz1")
    }

    private fun assertSuite(offset: Int, vararg labels: String) {
        val path = toPath(*labels)
        val element = getPsiElement(true, path.toString(), myFile)

        val expected = String.format("describe \"%s\"", labels[labels.size - 1])
        assertEquals(expected, firstLine(text(element)))
        assertEquals(offset, element.node.startOffset)
    }

    private fun assertTest(offset: Int, vararg labels: String) {
        val path = toPath(*labels)
        val element = getPsiElement(false, path.toString(), myFile)

        val expected = String.format("test \"%s\"", labels[labels.size - 1])
        assertEquals(expected, text(element))
        assertEquals(offset, element.node.startOffset)
    }

    private fun assertFuzz(offset: Int, vararg labels: String) {
        val path = toPath(*labels)
        val element = getPsiElement(false, path.toString(), myFile)

        val expected = String.format("fuzz fuzzer \"%s\"", labels[labels.size - 1])
        assertEquals(expected, text(element))
        assertEquals(offset, element.node.startOffset)
    }

    private fun assertMissing(vararg labels: String) {
        val path = toPath(*labels)
        val element = getPsiElement(false, path.toString(), myFile)
        assertSame(myFile, element)
    }

    private fun assertFallback(fallback: String, vararg labels: String) {
        val path = toPath(*labels)
        val element = getPsiElement(true, path.toString(), myFile)

        val expected = String.format("describe \"%s\"", fallback)
        assertEquals(expected, firstLine(text(element)))
    }

    private fun text(element: PsiElement): String {
        return element.text.trim { it <= ' ' }
    }

    private fun firstLine(text: String): String {
        return text.substringBefore("\n");
    }

}